test/framework/machinepool_helpers.go

```diff
@@ -139,6 +141,21 @@ func UpgradeMachinePoolAndWait(ctx context.Context, input UpgradeMachinePoolAndW
 
 		oldVersion := mp.Spec.Template.Spec.Version
 		mp.Spec.Template.Spec.Version = &input.UpgradeVersion
+
+		// This is a workaround for CAPD which has to use cgroupDriver: cgroupfs for kubernetes clusters < v1.24.
+		// It is required because kind and the node images do not support the default systemd for kubernetes < v1.24.
+		// This is to be removed as soon as no upgrades from kubernetes < v1.24 is tested anymore with CAPD.
+		if mp.Spec.Template.Spec.InfrastructureRef.Kind == "DockerMachinePool" {
+			if semver.MustParse(strings.TrimPrefix(input.UpgradeVersion, "v")).GTE(semver.MustParse("1.24.0")) {
+				// Ensure that we update to a BootstrapConfig which does not use cgroupfs
+				mp.Spec.Template.Spec.Bootstrap.ConfigRef.Name = strings.TrimSuffix(mp.Spec.Template.Spec.Bootstrap.ConfigRef.Name, "-cgroupfs")
+				// Reset DataSecretName so a new one gets created.
+				if strings.HasSuffix(*mp.Spec.Template.Spec.Bootstrap.DataSecretName, "-cgroupfs") {
```

Comment:
If I understand this correctly just changing the bootstrap ref in a MP will not be rolled out? If that's the case we should probably create an issue for it. I might be wrong but this sounds like a bug to me

Reply:
I'll recheck that

Reply:
My bad, looks like that was another issue while testing. I'll remove that part.

Reply:
Seems like I have to take another look, it breaks for the `pull-cluster-api-e2e-workload-upgrade-1-23-latest-main` test